### DISCUSSION TITLE
How to simulate Android KeyEvents?

### DISCUSSION THREAD

**Discussion by BrentWardOx** [wix/Detox#3363] [Q&A] [answered]
### Question

I'm developing a React Native app which listens for certain **_[KeyEvents](https://developer.android.com/reference/android/view/KeyEvent)_** on Android. I need to listen for KEYCODE_BACK, KEYCODE_MENU, KEYCODE_DPAD_UP, and a few others.

Pouring through the Detox documentation, I see that there is an Android-only method called **_[device.pressBack()](https://wix.github.io/Detox/docs/api/device-object-api#devicepressback-android-only)_** which specifically covers the KEYCODE_BACK event. That's great! My question is, how can I simulate **_other_** Android KeyEvents?

### What I've found so far

The biggest lead I've been able to find is **_[this post](https://github.com/wix/detox/issues/463#issuecomment-356320014)_** on a closed Detox issue from a few years ago. It is possible to simulate KeyEvents on attached Android devices/emulators with a terminal command like: `adb shell input keyevent 4` (4 is the keycode for the Android back key, can be substituted for any key code).

That command works great from a terminal window. Is is possible to execute that command from a Detox test file somehow?

**Comment by d4vidi** [comment, marked as answer]
Hi 😄
This is one of those cases where [UiAutomator's API](https://wix.github.io/Detox/docs/api/device-object-api/#devicegetuidevice-android-only) comes in handy, which is fully exposed through Detox. Good luck.

**Reply by BrentWardOx** [reply]
Thank you! I read about the UiAutomator in the Detox guide but did not realize it could be used for this purpose. I am able to use the UiAutomator to simulate KeyEvents like so:

```
const activeDevice = device.getUiDevice();
await activeDevice.pressKeyCode(19);
```

^ 19 is the KeyCode for DPAD_UP. Here is the full list of available **_[KeyCodes](https://developer.android.com/reference/android/view/KeyEvent.html)_**.
